src/dsymbol/tests.d

```diff
@@ -298,6 +298,81 @@ unittest
 	assert(T2.kind == CompletionKind.variadicTmpParam);
 }
 
+unittest
+{
+	writeln("Running public import tests...");
+
+	const dir = buildPath(tempDir(), "dsymbol");
+	const fnameA = buildPath(dir, "a.d");
+	const fnameB = buildPath(dir, "b.d");
+	const fnameC = buildPath(dir, "c.d");
+	const fnameD = buildPath(dir, "d.d");
+	const srcA = q{ int x; int w; };
+	const srcB = q{ float y; private float z; };
+	const srcC = q{ public { import a : x; import b; } import a : w; long t; };
+	const srcD = q{ public import c; };
+	// A simpler diagram:
+	// a = x w
+	// b = y [z]
+	// c = t + (x y) [w]
+	// d = (t x y)
+
+	mkdir(dir);
+	write(fnameA, srcA);
+	write(fnameB, srcB);
+	write(fnameC, srcC);
+	write(fnameD, srcD);
+	scope (exit)
+	{
+		remove(fnameA);
+		remove(fnameB);
+		remove(fnameC);
+		remove(fnameD);
+		rmdir(dir);
+	}
+
+	ModuleCache cache = ModuleCache(theAllocator);
+	cache.addImportPaths([dir]);
+
+	const a = cache.getModuleSymbol(istring(fnameA));
+	const b = cache.getModuleSymbol(istring(fnameB));
+	const c = cache.getModuleSymbol(istring(fnameC));
+	const d = cache.getModuleSymbol(istring(fnameD));
+	const ax = a.getFirstPartNamed(istring("x"));
+	const aw = a.getFirstPartNamed(istring("w"));
+	assert(ax);
+	assert(aw);
+	assert(ax.type && ax.type.name == "int");
+	assert(aw.type && aw.type.name == "int");
+	const by = b.getFirstPartNamed(istring("y"));
+	const bz = b.getFirstPartNamed(istring("z"));
+	assert(by);
+	assert(bz);
+	assert(by.type && by.type.name == "float");
+	assert(bz.type && bz.type.name == "float");
+	const ct = c.getFirstPartNamed(istring("t"));
+	const cw = c.getFirstPartNamed(istring("w"));
+	const cx = c.getFirstPartNamed(istring("x"));
+	const cy = c.getFirstPartNamed(istring("y"));
+	const cz = c.getFirstPartNamed(istring("z"));
+	assert(ct);
+	assert(ct.type && ct.type.name == "long");
+	assert(cw is null); // skipOver is true
+	assert(cx is ax);
+	assert(cy is by);
+	assert(cz is bz); // should not be there, but it is handled by DCD
```

Comment:
do you know why this happens or if it had previously happened before this change? I think fixing it in DCD is kind of the wrong approach for something like this.

Reply:
Because it does nothing. The library manages protection, but (almost) doesn't use it. I can write a fix for private symbols a bit later, in a separate PR. `getParts` method should have checks for imports importing private symbols.

Reply:
for imports and something like this this should definitely be done in dsymbol already. But I agree having a separate PR for it would be nicer then.

Reply:
So I've added protection checks, but they can filter out top-level symbols only, therefore we can't get rid of checks inside DCD.

I need this PR to be merged before another one (or I can just append it here, it is a fairly small change).